### ISSUE TITLE
Silent Hill 2 Arabic Language Pack

### ISSUE THREAD

**Issue by Adam-Saeed** [elishacloud/Silent-Hill-2-Enhancements#630]
Hi, I saw your [tweet](https://twitter.com/WhereAllBegins/status/1574559494633291795?s=20&t=S2R5si0K0FrI2l0zLMyTjw) about my Arabic language pack as posted on [Eternal Dream's Twitter](https://twitter.com/EternalDreamAR/status/1527542039730081792?s=20&t=Gsjkmfg4YTv_288fVOs3zg). I've originally made this in 2018 (before HD font) and re-released it with HD font in 2022. It would be my pleasure to host it on your site and maybe within your installer

The Arabic translation makes the following changes (and possibly their trouble):
- **the translation replaces the English message files** (if it won't causes issue, I'm ok with renaming them to another language)
- **Most of the message control codes are changed from left-aligning the text to center** (I hoped for right-alignment but it doesn't seem possible)
- **tga font** (replaces most of the Latin characters)
- **char table** (inside exe) I remember there's an issue discussing extracting it to a "fontwdata.bin" file
- **~~font setting in d3d8.ini~~** (I know, my work is horrible. But after writing this I've made further adjustments in char width to avoid using ini) but it was part of a messed up method to eliminate letter spacing by reducing the character width inside the executable down to the half or more (it was before LetterSpacing was an option); it creates a weird thing where the game doesn't crop the character and allows the characters to be closer to each other (it has been 3 years since I stumbled upon this and I still don't understand the logic behind it xD).
- **text inside exe isn't extracted**. there's "km\s" and "m" for the results screen inside exe that **was** translated but that executable was lost.
- **translated the map images** but couldn't enhance them and that created a bit of a mess. As you might know the map has a marker for the player positioning and some puzzles have a cursor and they're both in one file. What happens here is that the marker will disappear in either situation, if I remove the enhanced file "carsol.tex", the map marker would show up but the puzzle cursor would disappear and vise versa. So, I hope if it's possible to share with me the high-res version for the maps and I can re-translate them as the current texture tool can't extract them.

One last thing, is it possible to make HD font for SH3 and 4 or at least removing character spacing? It's the only reason keeping me from translating the rest of the series.

**Comment by Polymega** [comment]
Hi @Adam-Saeed,

We would be happy to host an Arabic translation pack on the site. The "Language Pack" page for the site will be featured whenever we release the next update. There's a few things we'll need to discuss and work out in order to have an Arabic translation pack in compliance with our project.

To start, I recommend you read over this guide, to help ensure your work follows our general guidelines:
https://docs.google.com/document/d/1X4k_2LFjS82ZTm1cbDRtUMB2x1qdfRjNIL0qUgPY7pg/edit#

More specifically, you will need to ensure your language pack follows these specific adjustments:
1. [Have you kept translations within a 4:3 boundary?](https://docs.google.com/document/d/1X4k_2LFjS82ZTm1cbDRtUMB2x1qdfRjNIL0qUgPY7pg/edit#heading=h.bgoye4s4iq8a)
2. A few lines in the game, such as "I'll wait at BAR Neely's" and "The door what wakes in darkness, opening into nightmares" should be blue, not white. [Have these colors been fixed?](https://docs.google.com/document/d/1X4k_2LFjS82ZTm1cbDRtUMB2x1qdfRjNIL0qUgPY7pg/edit#heading=h.1d337z371evw)
3. Nearly all of the save location names have their colors changed for the save menu, to make their coloring more consistent. [Have you changed the colors to be compliant?](https://docs.google.com/document/d/1X4k_2LFjS82ZTm1cbDRtUMB2x1qdfRjNIL0qUgPY7pg/edit#heading=h.ndiyy5lr2js)
4. There are character limitations for a few strings in the pause menu. [Are you within these character limitations?](https://docs.google.com/document/d/1X4k_2LFjS82ZTm1cbDRtUMB2x1qdfRjNIL0qUgPY7pg/edit#heading=h.fns41fbpz7b8)
5. Certain strings found within the Control Options menu can change, depending on whether Directional (2D) or Rotational (3D) control type is selected. [Have these strings been translated correctly?](https://docs.google.com/document/d/1X4k_2LFjS82ZTm1cbDRtUMB2x1qdfRjNIL0qUgPY7pg/edit#heading=h.r53nwexuz3m2)

^ I can help you check to ensure these things are included with your language pack at a later time.

There are a few other considerations for language packs to be hosted on our site, which might be a concern for an Arabic version:

- For the language packs that will be offered on the project site, we would not want to include a modified `sh2pc.exe` file if we can help it. Is there a way to create an Arabic language pack without modifying the main executable?
- The language packs offered on the site will only be for translating live text/subtitles. They will not include translated audio or image files. 
- You will probably want to replace another language besides English. This is because all non-English languages [include additional subtitles for events.](https://docs.google.com/document/d/1X4k_2LFjS82ZTm1cbDRtUMB2x1qdfRjNIL0qUgPY7pg/edit#heading=h.405in8lvfet3)

**For the HD font sheet:** The next project update will support PNG format for the HD font sheet. So you can save out the font sheet with a transparent background in PNG format. To test this, [you can use this build here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9701091/d3d8.zip). Be sure to name the font sheet `font000.png` and place it in the `sh2e\font\` folder.

Also note you can use [all characters in the full font sheet now](https://user-images.githubusercontent.com/27395260/186553032-82902734-f79c-430d-b23d-7b3f50b686e5.jpg), if needed. You might be already doing this for the Arabic language, but I wanted to let you know, just in case. If you have questions about the new, upcoming method of handling the font sheet, let us know.

**For the main menu text:** I can create a translated, HD main menu text for the Arabic language. I can't remember if I already made this for you or not? But if not, I can help you with this.

> One last thing, is it possible to make HD font for SH3 and 4 or at least removing character spacing? It's the only reason keeping me from translating the rest of the series.

The only person who might have done this is Belek666, who is also the creator of the SH2 HD font feature we're using here.

**Comment by Adam-Saeed** [comment]
I did read your document before writing this, I just forgot to point out what I haven't done yet.

> Have you kept translations within a 4:3 boundary?

Yes, we followed every line-break that was in the original text and never exceeded their position.

> A few lines in the game, such as "I'll wait at BAR Neely's" and "The door what wakes in darkness, opening into nightmares" should be blue, not white. Have these colors been fixed?

> Nearly all of the save location names have their colors changed for the save menu, to make their coloring more consistent. Have you changed the colors to be compliant?

No, I haven't done these yet, but I'm planning to

> There are character limitations for a few strings in the pause menu. Are you within these character limitations?

Yes, they were translated in the exe before by hex editing and I did copy them to the external file already.

> Certain strings found within the Control Options menu can change, depending on whether Directional (2D) or Rotational (3D) control type is selected. Have these strings been translated correctly?

Yes


> For the language packs that will be offered on the project site, we would not want to include a modified sh2pc.exe file if we can help it. Is there a way to create an Arabic language pack without modifying the main executable?

That's understandable, the only thing that is remaining inside the main executable is the font width, I just need to extract it to fontwdata.bin

> The language packs offered on the site will only be for translating live text/subtitles. They will not include translated audio or image files.

Sadly reading the map is important for the players to navigate the city.

> You will probably want to replace another language besides English. This is because all non-English languages include additional subtitles for events.

Would that be just as simple as just renaming the message files?

> Also note you can use [all characters in the full font sheet now](https://user-images.githubusercontent.com/27395260/186553032-82902734-f79c-430d-b23d-7b3f50b686e5.jpg), if needed. You might be already doing this for the Arabic language, but I wanted to let you know, just in case. If you have questions about the new, upcoming method of handling the font sheet, let us know.

I'm actually still using the short font sheet. Would jumping to the full sheet cause any character mapping issues? Also, is it ok if the font remains replacing Latin characters? I heard that anything past 200 is fixed width.

> For the main menu text: I can create a translated, HD main menu text for the Arabic language.

Oh, so the language would be available in the game's settings? I thought it would just be in the installer.

> I can't remember if I already made this for you or not? But if not, I can help you with this.

No, this is the first time I would post about Arabic translation here.

Thank you for your hard work

**Comment by Polymega** [comment]
> Would that be just as simple as just renaming the message files?

It would. Just change the "e" in `*_e.mes` to either "s", "i", "f", or "g".

> I'm actually still using the short font sheet. Would jumping to the full sheet cause any character mapping issues? Also, is it ok if the font remains replacing Latin characters?

No worries here. You're good to continue using the short font sheet if all you're replacing are the Latin characters. Our revised handling of the HD font sheet can accept short (partial) font sheets or full font sheets. We recommend saving it as a transparent PNG named `font000.png` for future use with our next project update. (You can use the test build I linked in the previous post to test this.)

> I heard that anything past 200 is fixed width.

That is correct. Only the Latin characters have variable width that can be adjusted. All Japanese characters are fixed-width with no means to adjust their width.

> Oh, so the language would be available in the game's settings? I thought it would just be in the installer.

I'm not totally sure what you're meaning, but the Arabic language would be a selectable language in the game's Options menu upon a user downloading the language pack (as it works for them now).

For the main menu (see below), I can translate this text to Arabic so that it can be included in your localization pack. (A translated main menu is the only texture file that'll be a part of language packs offered on the site.)

![image](https://user-images.githubusercontent.com/27395260/197357424-70c5833f-357b-4493-838a-6192f5fae107.png)

I will need the following translated to make this for you:

```
LOAD
CONTINUE
NEW GAME
OPTIONS
MOVIES
EXIT
ACTION LEVEL
RIDDLE LEVEL
HARD
NORMAL
EASY
BEGINNER
Main Scenario: LETTER FROM SILENT HEAVEN
Sub Scenario: BORN FROM A WISH
```

These language packs wouldn't be a part of the Setup Tool (installer), but rather additional download packs that can be attained through the project's website. Here's a WIP demo of how that will look/function on the site:

https://user-images.githubusercontent.com/27395260/195224814-257a127f-e846-470d-931e-203938eac6d9.mp4

**Comment by Adam-Saeed** [comment]
A little update. I managed to extract the font width data from the executable.

> I'm not totally sure what you're meaning, but the Arabic language would be a selectable language in the game's Options menu upon a user downloading the language pack (as it works for them now).

Yes, that's what I meant actually. That's awesome.

> I will need the following translated to make this for you:

Sure, but I don't know if it would show up correctly for you in your photo editing software. As for the font, it's on Google Fonts: https://fonts.google.com/specimen/Scheherazade+New
```
تحميل
مواصلة
لعبة جديدة
الخيارات
الأفلام
مغادرة
صعوبة الاشتباك
صعوبة الألغاز
صعب
متوسط
سهل
مبتدئ
السيناريو الرئيسي: رسالة من النعيم الصامت
السيناريو الجانبي: وليدة أمنية
```

> These language packs wouldn't be a part of the Setup Tool (installer), but rather additional download packs that can be attained through the project's website. Here's a WIP demo of how that will look/function on the site:

Can I add a link to my team's site or Twitter page?

> Sorry: You'd also need to change the language string in options_msg_*.mes.

I'll take that into consideration.

**Comment by Polymega** [comment]
Thanks for the translation! I will make this for you in the coming days.

> A little update. I managed to extract the font width data from the executable.

Awesome!

> Can I add a link to my team's site or Twitter page?

Absolutely. There will be a credits section for each language pack. I can add every person's name who was involved and their names can have links attached to them. Or, I can link to your team's website/Twitter page instead. Whatever you'd like to do here!

![image](https://user-images.githubusercontent.com/27395260/197361608-30466283-bfca-4964-a752-7c1ac491d0a6.png)

**Comment by Polymega** [comment]
@Adam-Saeed 
I have this mostly complete and ready for you:

![Screenshot 2022-10-23 13-46-20](https://user-images.githubusercontent.com/27395260/197407510-d6e19970-19e3-4bcf-97cb-eb6f77bb2ca5.png)
![Screenshot 2022-10-23 13-46-27](https://user-images.githubusercontent.com/27395260/197407515-3eb9de69-5d87-4d92-b84d-15594c864146.png)
![Screenshot 2022-10-23 13-46-33](https://user-images.githubusercontent.com/27395260/197407522-7f8cd328-1758-47f1-b57f-726c829d50f4.png)

However, there is overlap with two of the choices:

![image](https://user-images.githubusercontent.com/27395260/197407553-29c4b037-f3c2-4c72-8ee9-0256a49765f9.png)

Is it possible to use a different translation for "MOVIES"? Maybe "Videos" or "Scenes"?

**Comment by Adam-Saeed** [comment]
Awesome. it's looking great.

> Is it possible to use a different translation for "MOVIES"? Maybe "Videos" or "Scenes"?

Sure, I'll use "Videos"
```
الفيديوهات
```

**Comment by Polymega** [comment]
That's perfect, thanks. Here's that texture file. Change the "e" at the end of the filename if you have the Arabic language replace another language instead.
[start01e.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9855261/start01e.zip)

![Screenshot 2022-10-24 16-49-36](https://user-images.githubusercontent.com/27395260/197626810-85a92d85-2be4-4162-949a-76f81edc51e6.png)

**Comment by Adam-Saeed** [comment]
[sh2e Arabic Pack.zip](https://icedrive.net/s/hbfw6RFX8ika1uZSiA218ZRFF75w)
Sorry for the wait, I have made the required changes:
- The save location names have been standardized to what Silent Hill 2: Enhanced Edition uses
- "I'll wait at BAR Neely's" and "The door what wakes in darkness, opening into nightmares" have been colored correctly
- The ending letter is spaced-out to match the reading.
- Translation corrections
- This archive includes the font as png
- The translation is now replacing the German language

Also, since I replaced the Germen language, won't that leave the players with a jumbled mess of characters when they start the game with the default English language?

The font doesn't replace the capital letters, but I may re-do it to replace the Japanese characters later.

**Comment by Polymega** [comment]
> I have made the required changes:

Fantastic. Thank you so much.

> Also, since I replaced the Germen language, won't that leave the players with a jumbled mess of characters when they start the game with the default English language?

It won't, as the language pack will include a preconfigured `language.ini` file that will have the game launch with ~~German~~ Arabic already selected. In fact, I have re-packaged your files to be ready for release with the next project update. You can grab it here: [SH2EE_LangPack_Arabic.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/9894543/SH2EE_LangPack_Arabic.zip)

^ The only things I changed/added were including that preconfigured `language.ini` file and also renamed the `sh2e` folder to `lang`. Why was this folder changed to `lang`? This is to prevent your language pack from being overwritten whenever a user updates the project.

Previously, a user would have to place a language pack in the `sh2e` folder. The problem with doing this is that, whenever they updated something like the Enhanced Edition Essential Files package, it would overwrite/replace the language pack files. Now, we place language pack files in the `lang` folder as this folder will never be overwritten by project updates. This way, a user can download updates to the project without fear of the language pack breaking.

If you don't mind, please use the attached file/folder structure for use with the Arabic language pack. The next project update will utilize the `lang` folder, along with the launch of a "Language Pack" page on the project site.

It will be some time until we release this next project update, so if you make any further adjustments to your language pack during this time, just let me know and I'll update the files on my end accordingly.

Thanks, Adam-Saeed!

**Comment by Polymega** [comment]
Also, please tell me how you'd like the language pack to be credited on the project site, along with a link associated to each person's name/group name.

**Comment by Adam-Saeed** [comment]
AdamSaeed https://twitter.com/TheAdamSaeed
Venom Hussain

Sadly, I couldn't find any link for my friend.

**Comment by Polymega** [comment]
Thanks, Adam.

I also did some _very_ quick tests and noticed what seems to be an alignment issue with some of the text that present a selection:

![Screenshot 2022-10-29 22-07-42](https://user-images.githubusercontent.com/27395260/198859461-9f0f57bf-6ce2-43db-8263-cb65c4d48ce8.png)
![Screenshot 2022-10-29 22-09-20](https://user-images.githubusercontent.com/27395260/198859462-6fbf115b-2a49-4483-80d6-441657de3089.png)
![Screenshot 2022-10-29 22-11-16](https://user-images.githubusercontent.com/27395260/198859465-8fb901be-e0c2-4f99-91d8-9d8cdec2a230.png)

I only tested a few of these spots. This one looks to be centered correctly:

![Screenshot 2022-10-29 22-10-51](https://user-images.githubusercontent.com/27395260/198859473-15099f03-8e70-4156-8cee-f2f8890a800c.png)

**Comment by Adam-Saeed** [comment]
That's because there's a `<CENTER-PER-LINE>` in those lines. The last one showed up correctly because I removed it when revising some files. I forgot about the other instances.

**Comment by Polymega** [comment]
No worries. Just note I only tested this in a couple of spots, so you might want to do a full playthrough to be sure.

**Comment by Polymega** [comment]
Hi @Adam-Saeed,
Would you be able to fix the text alignment issues found in those spots where the game offers selection choices? (Seen a few posts above.)

I would recommend testing the whole game, as I only checked those few spots myselft.

**Comment by Adam-Saeed** [comment]
Yeah, it's fixable. I did test the whole game, too. In terms of text appearance issues, there was just one other issue beside this one. Though it requires re-importing all the message files. It's gonna take more time unfortunately.

**Comment by Polymega** [comment]
Thank you for looking into this and fixing it up. The Arabic community will be thrilled for your work.

**Comment by Polymega** [comment]
Hi Adam-Saeed,
Wanted to check in and see how the edits for the Arabic translation pack are going? If I can assist you in any way, let me know.
Thanks

**Comment by Polymega** [comment]
We only have a few things left before the launch of our next project update. If you'd like to include your Arabic translation pack as part of the next project update, please send me the packaged contents of your translation with the needed updates/revisions. Like before, I'm here to help and offer assistance, if needed.
Thanks

**Comment by Adam-Saeed** [comment]
Sorry, I just got caught up with many things. I appreaciate your offer to help but the translation requires redoing all the dialouge files which is sadly can only be done me. I will try to foucs on the bare minimun for the update but I can't really tell if it's managable.

**Comment by Polymega** [comment]
I'm pretty ignorant on what all needs to be done here, but I thought the translation was more-or-less finished? I thought the only thing needing attention [is fixing the center alignment issues for text strings that have choices on them](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/630#issuecomment-1296065272)?

**Comment by Adam-Saeed** [comment]
yeah, but during my testing I found grammatical error in the translation. It's not any normal language where I can open the dialouge files and review it there, it's a all latin characters that don't have any meaning (since the font replaces those characters with Arabic characters).

**Comment by Polymega** [comment]
@Adam-Saeed 
Checking in to see if everything is going okay with the translation revisions?

**Comment by Stoluos** [comment]
any news?
i can't wait to playing it with arabic subtitle :(

**Comment by Adam-Saeed** [comment]
Hello, I'm sorry. My life was unfortunately occupied by a number of events that impacted me. However, I'm still interested and want to provide the best possible quality

**Comment by Stoluos** [comment]
Finally, Adam, I came back
Praise be to God for your safety

**Comment by Stoluos** [comment]
i mean you not me

**Comment by Stoluos** [comment]
After waiting for your final return, Adam, the developer disappeared. Why is this happening? I am impatiently waiting for the Arabic language to be in the game. I hope you two will come back.

**Comment by Adam-Saeed** [comment]
Hi. I'm sorry for taking so long. I tried to proofread all the lines, but my perfectionism was a holding the project back for so long...
So, I did proofread a good number of them and I visited all the lines that has selection and fixed them.

[SH2EE_LangPack_Arabic_Fixed.zip](https://github.com/user-attachments/files/15506779/SH2EE_LangPack_Arabic_Fixed.zip)

**Comment by Adam-Saeed** [comment]
Forgot to mention, I contacted my friend VenomHussain and he said to use this link (https://x.com/_ar_team) for his credits.

So the whole credits would be:
AdamSaeed https://x.com/TheAdamSaeed
Venom Hussain https://x.com/_ar_team

**Comment by Polymega** [comment]
That's great news!!! I'll review the language pack as soon as I can to ensure nothing is missed. Do note that there have been a couple new project features added that will require you to make additional Arabic text strings. I will give you more information on this soon.

Thanks again!

**Comment by Adam-Saeed** [comment]
You're welcome, and I apologize for the disappearance. I'm interested in those new text strings and will be waiting for them.

**Comment by Polymega** [comment]
@Adam-Saeed 
Here are the new text strings that need to be added. You just need to reference the very first post at the top: https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/953#issue-2255321579

---------------------

I did a _very quick_ test with the translation pack and caught some errors. To start:

![Screenshot 2024-05-31 16-42-43](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/27395260/9bbf0f10-1429-4fa8-b791-42fb8ca3fd95)

^ I am unable to interact with this key, which softlocks the game as you cannot progress.

![Screenshot 2024-05-31 16-43-00](https://github.com/elishacloud/Silent-Hill-2-Enhancements/assets/27395260/f162ce0e-740c-4a77-9953-a0cfe5aeefd2)

^ When acquiring an item (such as the Handgun here), it's name ("Handgun") should be in green text.

-----------------------

After adding in those new strings, please take some time to go back and replay the game to ensure everything is behaving as expected.

Thanks

**Comment by Adam-Saeed** [comment]
It seems like the lines are being called from different files. The one about the key says, "I got a map of the west apt. bldg" and the one about the handgun says, "The stairs are broken. I can't go up any higher." This didn't happen before.

I think I know why this happened. Some files got corrupted. The file that had the line for the handgun is 0 byte. It seems the tool I'm using threw an error and I didn't notice. I have to work on a clean file

**Comment by Adam-Saeed** [comment]
@scigward Yeah, sadly things happened and had to leave pause the work on it. But I'm back now.

About the two images you included. These are new lines, I had just translated them. I noticed that they looked weird in file, but the new options didn't show up for me, so I couldn't test it out

**Comment by scigward** [comment]
Hey man just checking in, How is everything going about the Translation Pack?

**Comment by scigward** [comment]
Hey, Its been almost 3 months and you still gave no update, I'm just checking things out to see if everything is going okay, Thanks.
